deployer/server.go

```
@@ -0,0 +1,406 @@
+	if port == "" {
+		port = "4000"
+	}
+
+	addr := ":" + port
+	if err := discoverd.Register("flynn-deployer", addr); err != nil {
+		log.Fatal(err)
+	}
+
+	postgres.Wait("")
+	db, err = postgres.Open("", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := migrateDB(db.DB); err != nil {
+		log.Fatal(err)
+	}
+
+	pgxcfg, err := pgx.ParseURI(fmt.Sprintf("http://%s:%s@%s/%s", os.Getenv("PGUSER"), os.Getenv("PGPASSWORD"), db.Addr(), os.Getenv("PGDATABASE")))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	pgxpool, err := pgx.NewConnPool(pgx.ConnPoolConfig{
+		ConnConfig:   pgxcfg,
+		AfterConnect: que.PrepareStatements,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer pgxpool.Close()
+
+	q = que.NewClient(pgxpool)
+	wm := que.WorkMap{
+		"Deployment": handleJob,
+	}
+
+	// TODO start new worker on error
+	workers := que.NewWorkerPool(q, wm, 10)
+	go workers.Start()
+	defer workers.Shutdown()
+
+	router := httprouter.New()
+	router.POST("/deployments", createDeployment)
+	router.GET("/deployments/:deployment_id/events", streamDeploymentEvents)
+
+	log.Println("Listening for HTTP requests on", addr)
+	log.Fatal(http.ListenAndServe(addr, router))
+}
+
+func handleJob(job *que.Job) (e error) {
+	var args deployID
+	if err := json.Unmarshal(job.Args, &args); err != nil {
+		// TODO: log error
+		return err
+	}
+	id := args.ID
+	deployment, err := getDeployment(id)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	// for recovery purposes, fetch old formation
+	f, err := client.GetFormation(deployment.AppID, deployment.OldReleaseID)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	setDeploymentStatus(id, deployer.StatusRunning)
+	strategy, err := strategy.GetStrategy(deployment.Strategy, client)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	events := make(chan deployer.DeploymentEvent)
+	defer close(events)
+	go func() {
+		for ev := range events {
+			ev.DeploymentID = deployment.ID
+			sendDeploymentEvent(ev)
```

We should not ignore the error here (at the very least log it)